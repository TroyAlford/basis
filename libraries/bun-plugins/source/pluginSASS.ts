import type { BunPlugin } from 'bun'
import * as path from 'node:path'
import * as sass from 'sass'

/**
 * A Bun build plugin that compiles SASS/SCSS files into CSS and injects them into the document head.
 * The plugin automatically handles .scss file imports in your JavaScript/TypeScript code by compiling
 * the SASS content and creating a runtime script that injects the compiled CSS into the page.
 * @param root - The root directory path used to generate relative paths for debugging
 * @returns A Bun build plugin that handles SASS/SCSS compilation and injection
 * @example Basic usage in Bun build configuration
 * ```typescript
 * import { build } from "bun";
 * import { pluginSASS } from "./pluginSASS";
 *
 * await build({
 *   entrypoints: ['./src/index.ts'],
 *   plugins: [
 *     pluginSASS(process.cwd())
 *   ]
 * });
 * ```
 * @example Usage in your JavaScript/TypeScript files
 * // Will be compiled and injected at runtime
 * import './styles.scss'
 *
 * // Transformed to:
 * <style data-path="path/to/styles.scss">
 * // compiled CSS content
 * </style>
 */
export function pluginSASS(root: string): BunPlugin {
  return {
    name: 'scss',
    setup(build) {
      build.onLoad({ filter: /\.scss$/ }, async args => {
        const result = sass.compile(args.path)
        // Convert the SCSS import into a JavaScript module that exports the CSS
        return {
          contents: `
            const style = document.createElement('style');
            style.dataset.path = ${JSON.stringify(path.relative(root, args.path))};
            style.textContent = ${JSON.stringify(result.css)};
            document.head.appendChild(style);
          `,
          loader: 'js',
        }
      })
    },
  }
}
