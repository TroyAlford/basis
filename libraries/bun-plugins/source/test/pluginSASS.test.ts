import { describe, expect, test } from 'bun:test'
import * as path from 'node:path'
import { pluginSASS } from '../pluginSASS'

const FIXTURE_FILE = path.join(import.meta.dir, 'input.test.scss')
const JS_FILEPATH = './input.test.js'

describe('pluginSASS', () => {
  test.each([true, false])('minified: %p', async minify => {
    const build = await Bun.build({
      entrypoints: [FIXTURE_FILE],
      plugins: [pluginSASS({ minify })],
      target: 'browser',
    })

    expect(build.outputs).toHaveLength(1)
    expect(build.outputs.map(file => file.path)).toMatchSnapshot()
    const js = build.outputs.find(file => file.path === JS_FILEPATH)
    expect(await js.text()).toMatchSnapshot()
  })
})
