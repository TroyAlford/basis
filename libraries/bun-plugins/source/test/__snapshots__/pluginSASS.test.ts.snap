// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`pluginSASS minified: true 1`] = `
[
  "./input.test.js",
]
`;

exports[`pluginSASS minified: true 2`] = `
"// libraries/bun-plugins/source/test/input.test.scss
var style = document.createElement("style");
style.dataset.path = "libraries/bun-plugins/source/test/input.test.scss";
style.textContent = ".parent>.child{background-color:#333;padding:16px}";
document.head.appendChild(style);
"
`;

exports[`pluginSASS minified: false 1`] = `
[
  "./input.test.js",
]
`;

exports[`pluginSASS minified: false 2`] = `
"// libraries/bun-plugins/source/test/input.test.scss
var style = document.createElement("style");
style.dataset.path = "libraries/bun-plugins/source/test/input.test.scss";
style.textContent = \`.parent > .child {
  background-color: #333333;
  padding: 16px;
}\`;
document.head.appendChild(style);
"
`;
